nav2_behavior_tree/plugins/action/controller_cancel_node.cpp

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2022 Pradheep Padmanabhan - Neobotix GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <memory>
+
+#include "std_msgs/msg/string.hpp"
+
+#include "nav2_behavior_tree/plugins/action/controller_cancel_node.hpp"
+#include "lifecycle_msgs/msg/transition.hpp"
+
+namespace nav2_behavior_tree
+{
+
+using std::placeholders::_1;
+
+ControllerCancel::ControllerCancel(
+  const std::string & service_node_name,
+  const BT::NodeConfiguration & conf)
+: BtServiceNode<lifecycle_msgs::srv::ChangeState>(service_node_name, conf)
+{
+}
+
+void ControllerCancel::on_tick()
+{
+  request_->transition.id = lifecycle_msgs::msg::Transition::TRANSITION_ACTIVE_SHUTDOWN;
```

Comment:
> Why are you shutting down the server?! This should only be canceling the currently executing goal - how could this be transitioned back up if its deactivated for the next goal?

Shutting down will kill the entire control server. Once a new plan has been generated, the controller will kick start once the FollowPathAction BT Node kicks in again.

Reply:
> It should be canceling the action server's current goal, not shutting it down.

This is indeed a valid point, initially the problem was approached in this particular manner. But, just cancelling the goal (or deactivating) the action_server does not stop the robot from moving further. If you remember from the initial days when we started to talk about the issue, there was a point on "who sends the zero velocity?". Of course, we can extend the `controller_server` by adding the condition to `publishZeroVelocity`. But `shuttingdown` safely takes care of stopping the robot. In anycase we are not deallocating the configuration. 

Hope my argument makes sense!

Reply:
Deactivating the server is not an option, thats a misuse and its not correct that a new path will kick start the controller server once its deactivated. Some external stimulii will be required to transition the lifecycle state machine back into the activated state, which simply passing it a path will not accomplish

Reply:
I think this should just be calling the action server of the controller and cancelling any currently processing commands. 

if we cancel, it should stop the robot https://github.com/ros-planning/navigation2/blob/main/nav2_controller/src/controller_server.cpp#L379-L384

Reply:
> Deactivating the server is not an option, thats a misuse and its not correct that a new path will kick start the controller server once its deactivated.

Ah, I agree with your point completely now. The reason I was a bit adamant on this point was because, I was getting the intended behavior. I did a shutdown of the controller lifecycle node, but that does not mean we have unconfigured the controller as well right ? So, that means this the action_server_ is still active according to this [line here in the controller_server.](https://github.com/ros-planning/navigation2/blob/b201d058a7b15f1786c57afe69d079ac5d502018/nav2_controller/src/controller_server.cpp#L199). So, this indeed is a misuse.

Reply:
> Deactivating the server is not an option, thats a misuse and its not correct that a new path will kick start the controller server once its deactivated.

Ah, I agree with your point completely now. The reason I was a bit adamant on this point was because, I was getting the intended behavior. I did a shutdown of the controller lifecycle node, but that does not mean we have unconfigured the controller as well right ? So, that means the action_server_ is still active according to this [line here in the controller_server.](https://github.com/ros-planning/navigation2/blob/b201d058a7b15f1786c57afe69d079ac5d502018/nav2_controller/src/controller_server.cpp#L199). This indeed is a misuse. 

That is the reason why, the robot was able to still follow the new path and compute the necessary local plan and velocity command.

Reply:
I'm just in need of a little help in terms of how this BT Node needs to be structured using the action_server. The idea that I have now is to create a BT node derived from `BTActionNode` with the `FollowPath` action. Add the `asyncancelgoal()` logic to the `BTActionNode` and use that method from the `CancelControl` BT to cancel the action. This technically works. 

Now the issue is just, I'm not able to see how the test case should be. Because, there is no method in the controller_server or any other means apart from creating a new client in feeding the path to the controller_server for testing the BT Node. 

I also thought of having a separate `CancelAction` and integrate them to the controller_server. Even with this, again, there is no proper way to test it. 

Adding a `Path` parameter to `CancelControl` BT might solve the problem for testing, but on the other hand it does not make sense to have it just for the purpose of testing. 

Can I have some thoughts on here as well? That could be really helpful.

Reply:
You could create an action client (https://docs.ros2.org/dashing/api/rclcpp_action/classrclcpp__action_1_1Client.html) and use the `async_cancel_all_goals` or `async_cancel_goals_before` APIs to cancel goals being processed by the controller server altogether or before a specific time (e.g. before "now"). I'd recommend the latter just in case some stuff is happening quickly, don't want async communication error to get you down. I'd also remove a little bit from "now" so that its just slightly past dated, maybe 10ms. 

You can't unfortunately cancel a specific goal, which would be most ideal, except for in the location that has the actual goal handle that requested it, I believe. I don't think there's a way to construct a goal handle in another location. 

I don't think you need to derive from `BTActionNode`. There is probably some code from there you want to lift because its boilerplate you need, but I think you should just create your own node, it would be simpler. Else, you'd end up overriding like 75% of those functions and for what? Making your life harder. 

If anything, it may be useful to create a generic `BTCancelActionNode` such that we can create "cancel" nodes for controllers/planners/recoveries/etc from that basis, but that's another thing I don't think you need to do right now unless you would like to! :smile:

Reply:
> If anything, it may be useful to create a generic BTCancelActionNode such that we can create "cancel" nodes for controllers/planners/recoveries/etc from that basis, but that's another thing I don't think you need to do right now unless you would like to!

Scope of the project is getting much more clearer, I will indeed take up the additional work, since I think there is no urgency for this feature.

Reply:
> For testing BT nodes, you'd create a dummy action server you'd just name the controller server and call that instead.

This dummy server will be without any goal right? So, there will not be something to be cancelled.

Reply:
You can issue it a goal from the process and have the goal callback just be a while loop that is checking if canceled to exit.

Reply:
Cool, that's all clear now. BTW, I also created a `BTCancelActionNode`, if possible can I have a separate PR (also a new issue, with a list of possible "cancel" nodes if possible), by that way it will be easier for you to review that before and accordingly I can make changes to the nodes in this particular feature. 

Believe me, we are getting closer to the finish line.

Reply:
OK- submit the new PR and I can take a look at that alone! You can remove the implementation in this PR then